Add Playstation 4 Component

ktnrg45 opened the pull request:
## Description:
Add Playstation 4 component as media_player.ps4.

**Related issue (if applicable):** None

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7906

## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  platform: ps4
    host: 192.168.0.11
    region: 'R2'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

ktnrg45 committed:
Added Requirement for media_player.ps4

ktnrg45 committed:
Added media.player.ps4

ktnrg45 committed:
added component PS4 with ps4.py

homeassistant commented:
Hi @ktnrg45, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#19469).

Once you do that we will be able to review and accept this pull request.

Thanks!

ktnrg45 committed:
Shortened throttle time.

ktnrg45 committed:
Fixed error with referencing before assignment

ktnrg45 committed:
Fixed issue where name cannot be fetched correctly

ktnrg45 committed:
Fixed typo.

bryeartem commented:
I am getting this error when testing this component.. not really sure of the proper way to report this or help troubleshoot it but please let me know what I can do.

Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/core.py", line 1177, in _event_to_service_call
    await service_handler.func(service_call)
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/components/configurator.py", line 221, in async_handle_service_call
    call.data.get(ATTR_FIELDS, {}))
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/homeassistant/.homeassistant/custom_components/media_player/ps4.py", line 79, in configuration_callback
    result = credentials.listen()
  File "/home/homeassistant/.homeassistant/custom_components/media_player/ps4.py", line 452, in listen
    data = response[0]
UnboundLocalError: local variable 'response' referenced before assignment

ktnrg45 commented:
> I am getting this error when testing this component.. not really sure of the proper way to report this or help troubleshoot it but please let me know what I can do.
> 
> Traceback (most recent call last):
> File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/core.py", line 1177, in _event_to_service_call
> await service_handler.func(service_call)
> File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/components/configurator.py", line 221, in async_handle_service_call
> call.data.get(ATTR_FIELDS, {}))
> File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
> result = self.fn(*self.args, **self.kwargs)
> File "/home/homeassistant/.homeassistant/custom_components/media_player/ps4.py", line 79, in configuration_callback
> result = credentials.listen()
> File "/home/homeassistant/.homeassistant/custom_components/media_player/ps4.py", line 452, in listen
> data = response[0]
> UnboundLocalError: local variable 'response' referenced before assignment

Are you running you running Home Assistant or Hassio? If you're running regular home assistant  on an OS such as Debian/Raspbian, for example, you need to grant your python installation access to priveleged ports. 
To avoid running HA as sudo run command:
sudo setcap 'cap_net_bind_service=+ep' /usr/bin/python3.5 Replace "/usr/bin/python3.5" with your path to Python that is running Home Assistant.

See docs at:
https://github.com/home-assistant/home-assistant.io/blob/a6f884c7ea23facaec4708dbbdf9c1ae794a799a/source/_components/media_player.ps4.markdown

bryeartem commented:
That was it. running hassbian? (So many names) on Ubuntu works excellent now thanks so much! Awesome component appreciate the work!

ktnrg45 committed:
Code cleanup + Fixed Exceptions

ktnrg45 committed:
Added function to correct title if already stored

frenck committed:
Typo Fix

Co-Authored-By: ktnrg45 <38207570+ktnrg45@users.noreply.github.com>

ktnrg45 committed:
Typo fix

pvizeli commented:
We don't use `configuration` anymore. We replace it with `config_flow`. As an example, you can look to hue.

MartinHjelmare commented on homeassistant/components/media_player/ps4.py:
```diff
@@ -0,0 +1,510 @@
+                return
+
+
+class Dummy():
```
What's this used for?

If it's needed it has to be moved to a standalone library published on pypi. We don't want any device specific I/O code in home assistant.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

ktnrg45 replied:
See docstring right below line 399.
"""The PS4 Credentials object."""

ktnrg45 replied:
> https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services
> '#bad
> status = requests.get(url('/status'))'

Class does not call an API. Rather it  simply uses UDP sockets.

ktnrg45 replied:
Transferred to pypi package.

MartinHjelmare commented on homeassistant/components/media_player/ps4.py:
```diff
@@ -0,0 +1,510 @@
+            games.update(game)
+            self.save_games(games)
+
+    def get_ps_store_data(self, data_type, title):
```
This should be moved to a standalone library published on pypi.

ktnrg45 replied:
The function at 242? Please elaborate.

ktnrg45 replied:
Transfered both to imported package.

ktnrg45 committed:
Updated package requirement for ps4

ktnrg45 committed:
Moved art function and class to package

ktnrg45 commented:
> We don't use `configuration` anymore. We replace it with `config_flow`. As an example, you can look to hue.

@pvizeli I couldn't find any docs saying that `components/configurator.py` is currently or is planned to be deprecated @ https://www.home-assistant.io/components/configurator/ Could you post a reference?

simonk83 commented:
@ktnrg45 Presumably he's talking about this:  https://developers.home-assistant.io/docs/en/config_entries_config_flow_handler.html

kennedyshead commented on homeassistant/components/media_player/ps4.py:
```diff
@@ -0,0 +1,341 @@
+        """Retrieve the latest data."""
+        try:
+            status = self._ps4.get_status()
+            state = status.get('status')
```
There is no reson to store this in a local variable that I can see.

ktnrg45 replied:
> status = self._ps4.get_status()
This is an api call which communicates with a PS4 over LAN and returns a dict. Multiple states for the component are retrieved from that dict. My preference is to limit the # of api calls.

>  state = status.get('status')
Perhaps. If you would post an alternative, Ill look into it.

kennedyshead commented on homeassistant/components/media_player/ps4.py:
```diff
@@ -0,0 +1,341 @@
+                if titleid and name is not None:
+                    self._state = STATE_PLAYING
+                    self._media_content_id = titleid
+                    app_name = self._ps4.get_ps_store_data('title', name)
```
I would do this:
```python
app_name = self._ps4.get_ps_store_data(
    'title', name) or name
```

ktnrg45 replied:
Changes Made.

kennedyshead commented on homeassistant/components/media_player/ps4.py:
```diff
@@ -0,0 +1,341 @@
+
+    def load_games(self):
+        """Load games for sources."""
+        file = self._games_filename
```
imo you should never name a variable `file` (it shadows builtin) it should be `file_name´ or something

ktnrg45 replied:
Changes Made

kennedyshead commented on homeassistant/components/media_player/ps4.py:
```diff
@@ -0,0 +1,341 @@
+
+    def save_games(self, games):
+        """Save games to file."""
+        file = self._games_filename
```
see above

ktnrg45 replied:
Changes Made

kennedyshead commented on homeassistant/components/media_player/ps4.py:
```diff
@@ -0,0 +1,341 @@
+
+    def select_source(self, source):
+        """Select input source."""
+        if self._source_selected is not None:
```
So one can only select source one time? The only time it resets is if you manually turn off the game, is there any way around this?

ktnrg45 replied:
As of now. The PS4 prompts you to close current application. In testing I have found that you can switch sources multiple times.

ktnrg45 replied:
When source switches completely, the state should change to idle before booting game. Have you tried testing it?

kennedyshead replied:
Tested and it works nicely

kennedyshead reviewed (commented):
I have some suggestions

ktnrg45 committed:
Cleaned up code.

ktnrg45 committed:
Updated pypi lib.

ktnrg45 committed:
Updated pypi lib PS4

ktnrg45 committed:
Deleted commented line.

kennedyshead commented:
@ktnrg45 you have to run gen_requirements_all.py again

ktnrg45 committed:
Recompiled to latest.

ktnrg45 committed:
Corrected requirement_all

kennedyshead commented on requirements_all.txt:
```diff
@@ -1,5 +1,5 @@
 # Home Assistant core
```
You should only change the PS4 pypi right?

ktnrg45 committed:
Changed variable file.

ktnrg45 commented:
@MartinHjelmare Changes made. Ready for review.

kennedyshead commented:
Is this supposed to show media artwork? In that case it does not work on my setup

thibmaek commented:
I can't wait for this to land in HA core!

ktnrg45 committed:
Merge branch 'dev' into dev

ktnrg45 committed:
Pushed pyps4 lib to 0.1.4

ktnrg45 committed:
Pushed to 0.1.4, Reduced http requests in update()

ktnrg45 commented:
> Is this supposed to show media artwork? In that case it does not work on my setup

@kennedyshead Yes it should. I made a commit which may fix that. If it does not, please post the exact game title with edition type e.g.(Deluxe) if applicable which is not working with the link to the game in the Playstation Store website (https://store.playstation.com). Also if you can post the state data for your ps4 entity when game is running. 

FYI. I removed the source_selected code as I implemented a workaround.

ktnrg45 committed:
bumped pypi to 0.1.6

ktnrg45 committed:
bumped ps4 to 0.1.6

ktnrg45 committed:
Bump pyps4 0.1.7

ktnrg45 committed:
Bump to 0.1.7/ fix for update exception.

kennedyshead commented:
```bash
Traceback (most recent call last):
  File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/helpers/service.py", line 286, in _handle_service_platform_call
    await getattr(entity, func)(**data)
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/knutas/.homeassistant/custom_components/media_player/ps4.py", line 328, in select_source
    self._ps4.start_title(title_id)
  File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/pyps4_homeassistant/ps4.py", line 100, in start_title
    running_id = self.get_running_app_titleid()
  File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/pyps4_homeassistant/ps4.py", line 200, in get_running_app_titleid
    return self.get_status()['running-app-titleid']
KeyError: 'running-app-titleid'
```

ktnrg45 committed:
pylint; nested blocks

ktnrg45 committed:
Pylint cleaned update(), fixed souce_select()

ktnrg45 committed:
bump 0.1.8 pyps4

ktnrg45 committed:
whitespace

kennedyshead commented:
```
Traceback (most recent call last):
  File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/helpers/entity.py", line 221, in async_update_ha_state
    await self.async_device_update()
  File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/helpers/entity.py", line 349, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/util/__init__.py", line 315, in wrapper
    result = method(*args, **kwargs)
  File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/util/__init__.py", line 315, in wrapper
    result = method(*args, **kwargs)
  File "/home/knutas/.homeassistant/custom_components/media_player/ps4.py", line 178, in update
    name, titleid)
TypeError: cannot unpack non-iterable NoneType object
```

Minecraft https://store.playstation.com/en-se/product/EP4433-CUSA00265_00-MINECRAFTPS40000

maxandersen commented:
Trying this out running on hassio + lovelace and wondered if there a way to make it work there.

1) the homeassistant name never shows up in second screen app - I assume thats due to limitation of hass.io since if I run ps4walker locally in root mode I can do the pairing just fine.
   Proposed solution: Is there a way for user to write the config manually, i.e. use ps4-walker to generate the json ?

2) with lovelace the configuration step never shows up - have to go to classic ui ... I reckon this is because it uses the old configuration api ?

ktnrg45 commented:
> Trying this out running on hassio + lovelace and wondered if there a way to make it work there.
> 
> 1. the homeassistant name never shows up in second screen app - I assume thats due to limitation of hass.io since if I run ps4walker locally in root mode I can do the pairing just fine.
>    Proposed solution: Is there a way for user to write the config manually, i.e. use ps4-walker to generate the json ?
> 2. with lovelace the configuration step never shows up - have to go to classic ui ... I reckon this is because it uses the old configuration api ?

@maxandersen You have to start the config to have Home-Assistant show up in the 2nd Screen App. In the lovelace UI there is the alarm/bell icon at the top which shows notifications. The configurator should show up there.

Yes you can. Make a new file in your /config directory named ".ps4.conf" Inside it type your PS4 IP address and your credential code both in quotes ie. "192.168.0.14": "credential". Surround this in brackets {}

`{
"192.168.0.14": "credential"
}`

ktnrg45 commented:
> ```
> Traceback (most recent call last):
>   File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/helpers/entity.py", line 221, in async_update_ha_state
>     await self.async_device_update()
>   File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/helpers/entity.py", line 349, in async_device_update
>     await self.hass.async_add_executor_job(self.update)
>   File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
>     result = self.fn(*self.args, **self.kwargs)
>   File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/util/__init__.py", line 315, in wrapper
>     result = method(*args, **kwargs)
>   File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/util/__init__.py", line 315, in wrapper
>     result = method(*args, **kwargs)
>   File "/home/knutas/.homeassistant/custom_components/media_player/ps4.py", line 178, in update
>     name, titleid)
> TypeError: cannot unpack non-iterable NoneType object
> ```
> Minecraft https://store.playstation.com/en-se/product/EP4433-CUSA00265_00-MINECRAFTPS40000
> 
> State is playing

@kennedyshead Is there states for media_content_id?

maxandersen commented:
> Surround this in brackets {}. However as of now it won't let you pair to the PS4. I will be working on this soon.

is that something I can or need to do manually when running on hassio (since not running as root?) ?

ktnrg45 committed:
Renamed vars, split update() into functions

ktnrg45 committed:
typo fix

ktnrg45 commented:
> > Surround this in brackets {}. However as of now it won't let you pair to the PS4. I will be working on this soon.
> 
> is that something I can or need to do manually when running on hassio (since not running as root?) ?

No root required. Go to the same directory as your configuration.yaml and follow the instructions I posted earlier. FYI I have not tested this method. If you have already paired your ps4 with ps4-waker it may or may not work.

Have you tried looking for the configuration notification as I suggested earlier? I have tested this on Hassio/HassOS as well running as a custom component with no issues.

ktnrg45 commented:
> ```
> Traceback (most recent call last):
>   File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/helpers/entity.py", line 221, in async_update_ha_state
>     await self.async_device_update()
>   File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/helpers/entity.py", line 349, in async_device_update
>     await self.hass.async_add_executor_job(self.update)
>   File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
>     result = self.fn(*self.args, **self.kwargs)
>   File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/util/__init__.py", line 315, in wrapper
>     result = method(*args, **kwargs)
>   File "/home/knutas/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.85.0.dev0-py3.7.egg/homeassistant/util/__init__.py", line 315, in wrapper
>     result = method(*args, **kwargs)
>   File "/home/knutas/.homeassistant/custom_components/media_player/ps4.py", line 178, in update
>     name, titleid)
> TypeError: cannot unpack non-iterable NoneType object
> ```
> Minecraft https://store.playstation.com/en-se/product/EP4433-CUSA00265_00-MINECRAFTPS40000
> 
> State is playing

@kennedyshead You are in Sweden correct? The issue is that different Playstation regions have different sku's for games. Right now this only works for North America. I will work on a fix for this.

kennedyshead commented:
Aha! Well that explains it! Thank you for all your work.

ktnrg45 commented:
> > Have you tried looking for the configuration notification as I suggested earlier first? I have tested this on Hassio/HassOS 0.84 with running as a custom component with no issues.
> 
> yes, I open the configuration notification, used the secondary screen app but homeassistant never showed up.

Hmm. Are there any errors in your log? Are you using the official Hassio image, or Hassio installed on top of another OS?

ktnrg45 committed:
bumped ps4 0.2.1

ktnrg45 committed:
Added regions to configuration.

maxandersen commented:
> > yes, I open the configuration notification, used the secondary screen app but homeassistant never showed up.
> 
> Hmm. Are there any errors in your log? Are you using the official Hassio image, or Hassio installed on top of another OS?

Using official hassio. no, didn't see any errors/logging so I actually started putting loggng statements into see if it was actually doing something.

I can try restart  the setup with some more logging added and tell you what I see.

ktnrg45 commented:
> > > yes, I open the configuration notification, used the secondary screen app but homeassistant never showed up.
> > 
> > 
> > Hmm. Are there any errors in your log? Are you using the official Hassio image, or Hassio installed on top of another OS?
> 
> Using official hassio. no, didn't see any errors/logging so I actually started putting loggng statements into see if it was actually doing something.
> 
> I can try restart the setup with some more logging added and tell you what I see.

Silly question, but have you clicked the "start" button first in the config notification?

ktnrg45 commented:
> Aha! Well that explains it! Thank you for all your work.

No Problem. Thanks for the input. I committed a fix. Check if it works. You will have to add a line to your configuration.yaml:

media_player:
  platform: ps4
    host: 192.168.0.11
    region: 'R2'

kennedyshead commented:
I will test asap

maxandersen commented:
hmm - how do I trigger configuration to happen again ?

ktnrg45 commented:
> hmm - how do I trigger configuration to happen again ?

If the file `.ps4.conf` exists in your configuration directory, delete it. Then restart your Home-Assistant and it should show up under your notifications.

ktnrg45 commented:
Closing PR. Opening new PR due to number of changes made.
